File: photoeditor/src/main/java/ja/burhanrashid52/photoeditor/shape/RectangleShape.java
package ja.burhanrashid52.photoeditor.shape;

import android.graphics.Path;
import android.util.Log;

import androidx.annotation.NonNull;

public class RectangleShape extends AbstractShape {

    private float lastX, lastY;

    @Override
    protected String getTag() { return "RectangleShape"; }

    @Override
    public void startShape(float x, float y) {
        Log.d(getTag(), "startShape@ " + x + "," + y);
        left = x;
        top = y;
    }

    @Override
    public void moveShape(float x, float y) {
        right = x;
        bottom = y;

        float dx = Math.abs(x - lastX);
        float dy = Math.abs(y - lastY);
        if (dx >= TOUCH_TOLERANCE || dy >= TOUCH_TOLERANCE) {
            path = createRectanglePath();
            lastX = x;
            lastY = y;
        }
    }

    private @NonNull Path createRectanglePath() {
        Path path = new Path();
        path.moveTo(left, top);
        path.lineTo(left, bottom);
        path.lineTo(right, bottom);
        path.lineTo(right, top);
        path.close();
        return path;
    }

    @Override
    public void stopShape() {
        Log.d(getTag(), "stopShape");
    }

}

File: photoeditor/src/main/java/ja/burhanrashid52/photoeditor/shape/RectangleShape.kt
package ja.burhanrashid52.photoeditor.shape

import android.graphics.Path
import android.util.Log
import kotlin.math.abs

class RectangleShape : AbstractShape("RectangleShape") {
    private var lastX = 0f
    private var lastY = 0f

    override fun startShape(x: Float, y: Float) {
        Log.d(tag, "startShape@ $x,$y")
        left = x
        top = y
    }

    override fun moveShape(x: Float, y: Float) {
        right = x
        bottom = y
        val dx = abs(x - lastX)
        val dy = abs(y - lastY)
        if (dx >= TOUCH_TOLERANCE || dy >= TOUCH_TOLERANCE) {
            path = createRectanglePath()
            lastX = x
            lastY = y
        }
    }

    private fun createRectanglePath(): Path {
        val path = Path()
        path.moveTo(left, top)
        path.lineTo(left, bottom)
        path.lineTo(right, bottom)
        path.lineTo(right, top)
        path.close()
        return path
    }

    override fun stopShape() {
        Log.d(tag, "stopShape")
    }
}